extensions/amp-render/1.0/amp-render.js

```diff
@@ -0,0 +1,178 @@
+  constructor(element) {
+    super(element);
+
+    /** @private {?../../../src/service/template-impl.Templates} */
+    this.templates_ = null;
+
+    /** @private {?Element} */
+    this.template_ = null;
+  }
+
+  /** @override */
+  isLayoutSupported(layout) {
+    userAssert(
+      isExperimentOn(this.win, 'amp-render'),
+      'Experiment "amp-render" is not turned on.'
+    );
+    return super.isLayoutSupported(layout);
+  }
+
+  /** @override */
+  init() {
+    return dict({
+      'getJson': this.getJsonFn_(),
+    });
+  }
+
+  /**
+   * Returns the correct fetch function for amp-state, amp-script or
+   * to fetch remote JSON.
+   *
+   * @return {Function}
+   * @private
+   */
+  getJsonFn_() {
+    const src = this.element.getAttribute('src');
+    if (isAmpStateSrc(src)) {
+      return this.getAmpStateJson.bind(null, this.element);
+    }
+    if (isAmpScriptSrc(src)) {
+      // TODO(dmanek): implement this
+      return () => {};
+    }
+    return batchFetchJsonFor.bind(null, this.getAmpDoc(), this.element);
+  }
+
+  /**
+   * TODO: this implementation is identical to one in amp-data-display &
+   * amp-date-countdown. Move it to a common file and import it.
+   *
+   * @override
+   */
+  checkPropsPostMutations() {
+    const templates =
+      this.templates_ ||
+      (this.templates_ = Services.templatesForDoc(this.element));
+    const template = templates.maybeFindTemplate(this.element);
+    if (template != this.template_) {
+      this.template_ = template;
+      if (template) {
+        // Only overwrite `render` when template is ready to minimize FOUC.
+        templates.whenReady(template).then(() => {
+          if (template != this.template_) {
+            // A new template has been set while the old one was initializing.
+            return;
+          }
+          this.mutateProps(
+            dict({
+              'render': (data) => {
+                return templates
+                  .renderTemplateAsString(dev().assertElement(template), data)
+                  .then((html) => dict({'__html': html}));
+              },
+            })
+          );
+        });
+      } else {
+        this.mutateProps(dict({'render': null}));
+      }
+    }
+  }
+
+  /**
+   * TODO: this implementation is identical to one in amp-data-display &
+   * amp-date-countdown. Move it to a common file and import it.
+   *
+   * @override
+   */
+  isReady(props) {
+    if (this.template_ && !('render' in props)) {
+      // The template is specified, but not available yet.
+      return false;
+    }
+    return true;
+  }
+
+  /**
+   * Gets the json an amp-list that has an "amp-state:" uri. For example,
+   * src="amp-state:json.path".
+   *
+   * TODO: this implementation is identical to one in amp-list. Move it
+   * to a common file and import it.
+   *
+   * @param {!AmpElement} element
+   * @return {Promise<!JsonObject>}
+   */
+  getAmpStateJson(element) {
+    const src = element.getAttribute('src');
```

Comment:
Guard for if `src` is `null`, since we perform `slice` on it later. Also, let's make `src` a parameter, since it is called with that as an argument in `Render`.

Reply:
Added guard clause for when `src` is `null`. I'll add src as param in the follow up PR which will require moving `getAmpStateJson` out of `AmpRender` class  to support src binding

Reply:
Okay, in that case please add it as a TODO or in a tracking issue so we can keep track of it.